// @inrupt/vocab-common-rdf/GeneratedVocab/SCHEMA_INRUPT.ts
/**
 * MIT License
 * 
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the &quot;Software&quot;), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/**
 * Generated by the artifact generator [@inrupt/artifact-generator], version [1.0.1]
 * as part of artifact: [vocab-common-rdf], version: [1.0.1]
 * on 'Friday, August 20, 2021 4:35 PM'.
 *
 * Vocabulary built from vocab list file: [common-rdf/vocab-common-rdf.yml].
 *
 * Inrupt extension to Schema.org terms providing
 multilingual alternative names (i.e. labels) and translations for comments
 (e.g. for use directly as labels or tool-tips in user interfaces or error
 messages). This extension very deliberately cherry-picks the individual terms
 from Schema.org that Inrupt currently deem generally useful for Solid and Solid
 applications (meaning we can provide a much cleaner, less noisy and smaller
 bundle size when generating programming language artifacts that provide
 convenient constants for just these selected terms, rather than including the
 over 2,500 terms currently defined in Schema.org).
 */

// We prefix our local variables with underscores to (hopefully!) prevent
// potential names clashes with terms from vocabularies.
const _NAMESPACE = "http://schema.org/";
function _NS (localName: string): string {
  return (_NAMESPACE + localName);
}

/**
 * Inrupt extension to Schema.org terms providing
 multilingual alternative names (i.e. labels) and translations for comments
 (e.g. for use directly as labels or tool-tips in user interfaces or error
 messages). This extension very deliberately cherry-picks the individual terms
 from Schema.org that Inrupt currently deem generally useful for Solid and Solid
 applications (meaning we can provide a much cleaner, less noisy and smaller
 bundle size when generating programming language artifacts that provide
 convenient constants for just these selected terms, rather than including the
 over 2,500 terms currently defined in Schema.org).
 */
const SCHEMA_INRUPT = {
  PREFIX: "schema-inrupt",
  NAMESPACE: _NAMESPACE,
  PREFIX_AND_NAMESPACE: { "schema-inrupt": "http://schema.org/" },
  NS: _NS,

  // *****************
  // All the Classes.
  // *****************

  /**
   * A person (alive, dead, undead, or fictional).
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  Person: _NS("Person"),

  /**
   * Data type: URL.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  URL: _NS("URL"),

  /**
   * Any medical intervention designed to prevent, treat, and cure human diseases and medical conditions, including both curative and palliative therapies. Medical therapies are typically processes of care relying upon pharmacotherapy, behavioral therapy, supportive therapy (with fluid or nutrition for example), or detoxification (e.g. hemodialysis) aimed at improving or preventing a health condition.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  MedicalTherapy: _NS("MedicalTherapy"),

  /**
   * An educational or occupational credential. A diploma, academic degree, certification, qualification, badge, etc., that may be awarded to a person or other entity that meets the requirements defined by the credentialer.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  EducationalOccupationalCredential: _NS("EducationalOccupationalCredential"),


  // *******************
  // All the Properties.
  // *******************

  /**
   * An alias for the item.
   *
   * This term has [6] labels and comments, in the languages [NoLocale, de, en, es, fr, it].
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  alternateName: _NS("alternateName"),

  /**
   * Given name. In the U.S., the first name of a Person.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  givenName: _NS("givenName"),

  /**
   * Family name. In the U.S., the last name of a Person.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  familyName: _NS("familyName"),

  /**
   * An additional name for a Person, can be used for a middle name.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  additionalName: _NS("additionalName"),

  /**
   * A license document that applies to this content, typically indicated by URL.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  license: _NS("license"),

  /**
   * The name of the item.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  name: _NS("name"),

  /**
   * The textual content of this CreativeWork.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  text: _NS("text"),

  /**
   * The identifier property represents any kind of identifier for any kind of [[Thing]], such as ISBNs, GTIN codes, UUIDs etc. Schema.org provides dedicated properties for representing many of these, either as textual strings or as URL (URI) links. See [background notes](/docs/datamodel.html#identifierBg) for more details.
        
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * See also:
   *  - https://schema.org/docs/datamodel.html#identifierBg
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  identifier: _NS("identifier"),

  /**
   * A description of the item.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  description: _NS("description"),

  /**
   * An image of the item. This can be a [[URL]] or a fully described [[ImageObject]].
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * See also:
   *  - https://schema.org/ImageObject
   *  - https://schema.org/URL
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  image: _NS("image"),

  /**
   * URL of the item.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  url: _NS("url"),

  /**
   * The startTime of something. For a reserved event or service (e.g. FoodEstablishmentReservation), the time that it is expected to start. For actions that span a period of time, when the action was performed. e.g. John wrote a book from *January* to December. For media, including audio and video, it's the time offset of the start of a clip within a larger file.\n\nNote that Event uses startDate/endDate instead of startTime/endTime, even when describing dates with times. This situation may be clarified in future revisions.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  startTime: _NS("startTime"),

  /**
   * The endTime of something. For a reserved event or service (e.g. FoodEstablishmentReservation), the time that it is expected to end. For actions that span a period of time, when the action was performed. e.g. John wrote a book from January to *December*. For media, including audio and video, it's the time offset of the end of a clip within a larger file.\n\nNote that Event uses startDate/endDate instead of startTime/endTime, even when describing dates with times. This situation may be clarified in future revisions.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  endTime: _NS("endTime"),

  /**
   * The start date and time of the item (in [ISO 8601 date format](http://en.wikipedia.org/wiki/ISO_8601)).
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  startDate: _NS("startDate"),

  /**
   * The end date and time of the item (in [ISO 8601 date format](http://en.wikipedia.org/wiki/ISO_8601)).
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  endDate: _NS("endDate"),

  /**
   * Email address.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  email: _NS("email"),

  /**
   * A preventative therapy used to prevent an initial occurrence of the medical condition, such as vaccination.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  primaryPrevention: _NS("primaryPrevention"),

  /**
   * Specific qualifications required for this role or Occupation.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  qualifications: _NS("qualifications"),

  /**
   * A person or organization attending the event.
   *
   * This term provides multilingual descriptions, with [6] labels in languages [NoLocale, de, en, es, fr, it], but [5] comments in languages [NoLocale, de, es, fr, it] (so the difference is only between English and NoLocale, which we consider the same).
   *
   * Defined by the vocabulary: https://w3id.org/inrupt/vocab/extension/schema#
   */
  attendee: _NS("attendee"),

}

export default SCHEMA_INRUPT;
